### ISSUE TITLE
KeyOf schema change in 24.0

### ISSUE THREAD

**Issue by zedeus** [sinclairzx81/typebox#188]
Enum strings are excellent in OpenAPI. The usage of `allOf` for enums and unions is messy and hard to read in documentation generated by most OpenAPI/Swagger tooling. The solution in #175 is not sufficient since adding `$id` doesn't work from what I can tell, and it's just generally broken (note how the generated schema is empty):
![image](https://user-images.githubusercontent.com/9434921/176668260-1eb9dfa9-7164-4084-ae22-0026788a5f15.png)
```ts
type IntoStringUnion<T> = {
	[K in keyof T]: T[K] extends string ? TLiteral<T[K]> : never;
};

export function StringUnion<T extends string[]>(
	values: [...T]
): TUnion<IntoStringUnion<T>> {
	// eslint-disable-next-line @typescript-eslint/no-unsafe-return
	return { enum: values } as any;
}

export const Role = StringUnion(["internal", "owner", "admin", "communicator"]);
export type Role = Static<typeof Role>;

export const Roles = T.Array(Role, {
	minItems: 1,
	maxItems: 1,
	$id: "Roles"
});
```

As a workaround I used `KeyOf` to generate a proper string enum, but now it spits out the messy `allOf` object again:
```ts
export function StringUnion<T extends string>(
	values: T[],
	options?: SchemaOptions
) {
	const obj = Object.fromEntries(
		values.map((k) => [k, Type.String()])
	) as Record<T, TString>;

	return Type.KeyOf(Type.Object(obj), options);
}
```

23.4:
https://github.com/sinclairzx81/typebox/blob/17c255319d044fb99b9b770dfa5f2b9f67e06924/src/typebox.ts#L382-L387
![image](https://user-images.githubusercontent.com/9434921/176665152-29300f6d-7a5f-4e36-8d9a-1c666dd3f78b.png)

24.1:
https://github.com/sinclairzx81/typebox/blob/4911a8f30c81607de1e5d26183b98947cc4e877c/src/typebox.ts#L619-L623
![image](https://user-images.githubusercontent.com/9434921/176666312-a3d3c53c-3003-458d-b2a7-98dccff81637.png)

New solution:
```ts
export function StringUnion<T extends string>(
	values: T[],
	options: SchemaOptions
) {
	return Type.Unsafe<T>({ ...options, type: "string", enum: values });
}
```

**Comment by sinclairzx81** [comment]
@zedeus Hi. As of 0.24.0 by default, TypeBox will standardize on the `anyOf` representation for all things union-like. The reason for this stems from the need to enable more composition scenarios for end users, and selecting one form of schema representation enables TypeBox to rely on that representation to enable things like the following.

```typescript
const T = Type.Object({
  x: Type.Number(),
  y: Type.Number()
})

const K = Type.KeyOf(T)                  

const R = Type.Record(K, Type.String())  // Record<'x' | 'y', string>
```

As you've noticed though, the new `Type.Unsafe()` function was released with 0.24.0, and provides a easier means to construct arbitrary schemas (including `enum` representations).

```typescript
import { Type, Static, TObject } from '@sinclair/typebox'

// -------------------------------------------------------------
// EnumKeyOf<T>
// -------------------------------------------------------------

export function EnumKeyOf<T extends TObject>(schema: T) {
    return Type.Unsafe<keyof T['properties']>({ enum: Object.keys(schema.properties) })
}

const K = EnumKeyOf(Type.Object({
    a: Type.Number(),
    b: Type.Number()
}))

type K = Static<typeof K> // type K = 'a' | 'b'

// -------------------------------------------------------------
// StringEnum<T>
// -------------------------------------------------------------

export function StringEnum<T extends string[]>(_enum: [...T]) {
    return Type.Unsafe<T[number]>({ enum: _enum })
}

const E = StringEnum(['A', 'B', 'C'])

type E = Static<typeof E> // type E = 'A' | 'B' | 'C'
```

Does this help to resolve your use case?
Let me know
S

**Comment by zedeus** [comment]
I missed the Unsafe part in the readme, my bad. Thank you for the detailed response!
